File: python/strings/msd_radix_sort.py
"""MSD (most significant digit) string sort algorithm.
This sorting algorithm works sorting strings character by character
from left to right.

This is a different implementation than the one found in the
file msd_string_sort.py. It has less LOC because it makes use
of list comprehensions. Also, instead of just counting the frequencies,
it keeps the whole strings inside 'buckets'.
"""

# R = radix. The number of characters of the alphabet supported
# by this implementation of Radix Sort. It won't work with Unicode.
R = 256  # -> Extended ASCII.


def msd_radix_sort(data, digit=0):
    """Recursively sort a list of strings. Strings are sorted char by char
    from left to right. `digit` marks the current char by which their are
    going to be sorted.
    :param data: list of strings to be sorted.
    :param digit: string position by which strings are going to be sorted.
    :return: list of strings
    """

    # print("\t\t" * digit, "->", data, digit)

    # If the list contains just one string, return it.
    if len(data) <= 1:
        return data

    # Create a bucket for every possible character.
    buckets = [[] for _ in range(R)]
    complete_bucket = []

    # Locate every string in a bucket according to the character of it
    # that is being processed (`digit` points to a position in the string).
    for s in data:
        if len(s) > digit:
            buckets[ord(s[digit])].append(s)
        else:
            complete_bucket.append(s)

    # For every bucket, run radix sort on the next digit.
    buckets = [msd_radix_sort(buckets[r], digit + 1) for r in range(R) if len(buckets[r]) > 0]

    # Flatten `buckets` and concat with `complete_bucket`.
    return complete_bucket + [bucket for bucket_list in buckets for bucket in bucket_list]


if __name__ == "__main__":
    unsorted_data = [
        "b",
        "ab",
        "ca",
        "aaa",
        "c"
    ]
    print("-> ", msd_radix_sort(unsorted_data))

File: javascript/strings/msd-radix-sort.js
/** MSD (most significant digit) string sort algorithm.
This sorting algorithm works sorting strings character by character
from left to right.

Characteristics:
    - Stable (preserves original order).

Main differences with the LSD string sort algorithm:
    - Runs from left to right.
    - It's recursive.
*/

"use strict";

// Radix: number of different possible characters.
// 256 -> Extended ASCII.
const R = 256;

function initializeArray(size, defaultValue) {
  var a = [];
  for (var i = 0; i < size; i++) a.push(defaultValue);
  return a;
}

function charCode(str, digit) {
  if (digit < str.length) {
    return str.charCodeAt(digit);
  } else {
    return -1;
  }
}

function MSDRadixSort(data) {
  var sortedData = [...data];
  var aux = initializeArray(data.length, undefined);
  msdRadixSort(sortedData, aux, 0, data.length - 1, 0);
  return sortedData;
}

function msdRadixSort(data, aux, low, high, digit) {
  if (high <= low) return;

  // R + 2 because we need one more space to accomodate the -1 case.
  var count = initializeArray(R + 2, 0);

  // Phase 1: compute frequencies.
  for (var i = low; i <= high; i++) {
    count[charCode(data[i], digit) + 2]++;
  }

  // Phase 2: compute indices (starting points for each letter/character).
  for (var r = 0; r <= R; r++) {
    count[r + 1] += count[r];
  }

  // Phase 3: distribute data.
  for (var i = low; i <= high; i++) {
    var newPosition = count[charCode(data[i], digit) + 1]++;
    aux[newPosition] = data[i];
  }

  // Phase 4: copy sorted data into data.
  for (var i = low; i <= high; i++) {
    data[i] = aux[i - low];
  }

  // Sort every partition
  for (var r = 0; r < R; r++) {
    msdRadixSort(data, aux, low + count[r], low + count[r + 1] - 1, digit + 1);
  }
}

module.exports = {
  MSDRadixSort
};
